time/src/time.rs

```diff
@@ -29,21 +29,152 @@ pub(crate) enum Padding {
+    padding: Padding,
+
+    // Big endian version
+    #[cfg(target_endian = "big")]
+    #[allow(clippy::missing_docs_in_private_items)]
+    hour: u8,
+    #[cfg(target_endian = "big")]
     #[allow(clippy::missing_docs_in_private_items)]
     minute: u8,
+    #[cfg(target_endian = "big")]
     #[allow(clippy::missing_docs_in_private_items)]
     second: u8,
+    #[cfg(target_endian = "big")]
     #[allow(clippy::missing_docs_in_private_items)]
     nanosecond: u32,
+    #[cfg(target_endian = "big")]
     #[allow(clippy::missing_docs_in_private_items)]
     padding: Padding,
 }
 
+impl core::hash::Hash for Time {
+    fn hash<H: core::hash::Hasher>(&self, state: &mut H) {
+        self.as_u64().hash(state)
+    }
+}
+
+impl PartialEq for Time {
+    fn eq(&self, other: &Self) -> bool {
+        self.as_u64().eq(&other.as_u64())
+    }
+
+    #[allow(clippy::partialeq_ne_impl)]
+    fn ne(&self, other: &Self) -> bool {
+        self.as_u64().ne(&other.as_u64())
+    }
+}
+
+impl Eq for Time {}
+
+impl PartialOrd for Time {
+    fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+        self.as_u64().partial_cmp(&other.as_u64())
+    }
+
+    fn lt(&self, other: &Self) -> bool {
+        self.as_u64().lt(&other.as_u64())
+    }
+
+    fn le(&self, other: &Self) -> bool {
+        self.as_u64().le(&other.as_u64())
+    }
+
+    fn gt(&self, other: &Self) -> bool {
+        self.as_u64().gt(&other.as_u64())
+    }
+
+    fn ge(&self, other: &Self) -> bool {
+        self.as_u64().ge(&other.as_u64())
+    }
+}
+
+impl Ord for Time {
+    fn cmp(&self, other: &Self) -> core::cmp::Ordering {
+        self.as_u64().cmp(&other.as_u64())
+    }
+
+    fn max(self, other: Self) -> Self
+    where
+        Self: Sized,
+    {
+        let max = self.as_u64().max(other.as_u64());
+        // SAFETY:
+        // This is safe because max is the result of
+        // Time::as_u64
+        unsafe { Self::from_u64(max) }
+    }
+
+    fn min(self, other: Self) -> Self
+    where
+        Self: Sized,
+    {
+        let min = self.as_u64().min(other.as_u64());
+        // SAFETY:
+        // This is safe because min is the result of
+        // Time::as_u64
+        unsafe { Self::from_u64(min) }
+    }
+
+    fn clamp(self, min: Self, max: Self) -> Self
+    where
+        Self: Sized,
+    {
+        let clamp = self.as_u64().clamp(min.as_u64(), max.as_u64());
+        // SAFETY:
+        // This is safe because clamp is the result of
+        // Time::as_u64
+        unsafe { Self::from_u64(clamp) }
+    }
+}
+
 impl Time {
+    /// Provides an u64 based representation **of the correct endianness**
+    ///
+    /// This representation can be used to do comparisons equality testing or hashing.
+    const fn as_u64(self) -> u64 {
+        // SAFETY:
+        // Transmuting to an u64 is always safe
+        // given that size is checked by transmute
+        // and alignment is irrelevant.
+        unsafe { core::mem::transmute(self) }
```

Comment:
Why is `unsafe` necessary here? It should be possible to get the same result using `u64::from_le_bytes` or similar, I would imagine.

Reply:
IIRC with `u64::from_le_bytes` the compiler didn't manage to see this was a transmute. And as above, I felt it was better to be more explicit rather than rely on compiler optimization.

Reply:
Relying on compiler optimizations is fine. Some things in `time` already do. It's not as if the code will be broken if it's not as optimized.